### PR TITLE
[FLOC 2087] Example: Using the API with Python

### PR DESCRIPTION
**Pull request by davidoliverSP2**
These examples (this being the first of 3 currently under development) are currently tested manually in the following repo: https://github.com/ClusterHQ/flocker-api-examples

The README.md file details the process of testing, and the results of the tests can be viewed in the "_output" folder.

The process at the moment is that I run the test manually before copying the example files into the Flocker repo. We have an issue to bring these tests into the automated Buildbot tests post v1. See FLOC 2395.

When designing this example, @itamarst recommended that a good way to do this was using requests libraries. Having discussed this with @binocarlos, we decided the example "as-is" is affective, but we opened an issue to create an improved example using libraries post v1.0. See FLOC 2377.

### CODE REVIEW

File: docs/using/examples/api-example-python.rst [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+#. Download the :file:`example.py` file:
+
+   :download:`example.py`
+
+   .. literalinclude:: example.py
+
+#. Before you run the example, you will need to change the following values (either by editing the source file or by setting environment variables):
+
+   * **CONTROL_SERVICE** - the IP address of the Flocker control service
```
Comment (adamtheturtle) [inline docs/using/examples/api-example-python.rst]: Sphinx has ``:envvar:`` http://sphinx-doc.org/markup/inline.html#role-envvar.

**Comment by adamtheturtle** [comment]
This is marked as a task in progress in JIRA, is it meant to be an improvement in code review ready?

File: docs/using/examples/example.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+import httplib
+import os
+import ssl
+import tempfile
+
+# NOTE - to run this example, you need at least python 2.7.9
```
Comment (adamtheturtle) [inline docs/using/examples/example.py]: :question: Maybe instead of noting this, let the code error. Or make it error with a "python version not sufficient" type error.

File: docs/using/examples/api-examples.rst [hunk 3]
```diff
@@ -0,0 +1,14 @@
+:tocdepth: 2
+
+====================
+Flocker API Examples
+====================
+
+You can find below an example of using the :ref:`Flocker REST API<api>` with Python.
+
+Examples of using the Flocker API with languages such as `cURL`, `Go`, and `Ruby` will be available soon.
```
Comment (adamtheturtle) [inline docs/using/examples/api-examples.rst]: Single backticks are for italics, I think that double backticks are what you want.

Is it useful to say what will be available soon? It adds risk in case the roadmap changes (and this is software, it does :smile: ).

File: docs/using/examples/example.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+# Define the control IP, port, and the certificates for authentication.
+
+CONTROL_SERVICE = os.environ.get(
+    "CONTROL_SERVICE", "54.157.8.57")
+CONTROL_PORT = os.environ.get(
+    "CONTROL_PORT", 4523)
+KEY_FILE = os.environ.get(
+    "KEY_FILE", "/Users/kai/projects/flocker-api-examples/flockerdemo.key")
```
Comment (adamtheturtle) [inline docs/using/examples/example.py]: Hold on while I create /Users/kai :laughing:

File: docs/using/examples/api-example-python.rst [hunk 5]
```diff
@@ -0,0 +1,27 @@
+=========================
+Using the API with Python
+=========================
+
+.. note:: You will need to have a running Flocker cluster set up, and at least `Python 2.7.9`_ to run this example.
+
+#. Download the :file:`example.py` file:
```
Comment (adamtheturtle) [inline docs/using/examples/api-example-python.rst]: Name this file something more informative.

File: docs/using/examples/api-example-python.rst [hunk 6]
```diff
@@ -0,0 +1,27 @@
+=========================
+Using the API with Python
+=========================
+
+.. note:: You will need to have a running Flocker cluster set up, and at least `Python 2.7.9`_ to run this example.
```
Comment (adamtheturtle) [inline docs/using/examples/api-example-python.rst]: Python 3 is "at least Python 2.7.9", right?

Comment (adamtheturtle) [inline docs/using/examples/api-example-python.rst]: How do I set up a Flocker cluster (link I guess).

File: docs/using/examples/api-example-python.rst [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+#. Download the :file:`example.py` file:
+
+   :download:`example.py`
+
+   .. literalinclude:: example.py
+
+#. Before you run the example, you will need to change the following values (either by editing the source file or by setting environment variables):
```
Comment (adamtheturtle) [inline docs/using/examples/api-example-python.rst]: The prerequisites are oddly split between here and the note above.

File: docs/using/examples/example.py (hunk 4 above)
Comment (adamtheturtle) [inline docs/using/examples/example.py]: Ah, I see that a note says to change this. If we really want example users, let's stick with `alice` for consistency.

File: docs/using/examples/example.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+# Make the first request to check the service is working.
+
+make_api_request("GET", "/v1/version")
+
+# Create a volume.
+
+make_api_request("POST", "/v1/configuration/datasets",
+    data= r'{"primary": "%s", "maximum_size": 107374182400, "metadata": {"name": "example_dataset"}}'
+        % ("5540d6e3-392b-4da0-828a-34b724c5bb80",))
```
Comment (adamtheturtle) [inline docs/using/examples/example.py]: Newline at EOF

**Comment by adamtheturtle** [comment]
> We have an issue to bring these tests into the automated Buildbot tests post v1

Now is post v1 :smile: .

Adding untested examples has problems including:
* If they stop working because of a Flocker change, we might miss that change, and then the examples will break. For example if we change an environment variable name, or add a new required one. We increase the risk of having non-working examples in our docs.
* If they stop working because of a non-Flocker change (in this case unlikely, but e.g. if we used a library and the library's API changed), we might not know. We increase the risk of having non-working examples in our docs.
* Every reviewer (read: every existing and new employee with access to merge on GitHub) has to consider these examples for every API change. That overhead slows down development.
* If we make a change to the API and we think "huh, maybe this stops the API example working" the testing of that is slow and manual.
* If we want to improve this example and test that it still works then:
    * That is slow and manual.
    * We might miss an edge case which the original author considered, but isn't documented because maybe they thought that it was "obvious".

File: docs/using/examples/api-examples.rst [hunk 9]
```diff
@@ -0,0 +1,14 @@
+:tocdepth: 2
+
+====================
+Flocker API Examples
```
Comment (adamtheturtle) [inline docs/using/examples/api-examples.rst]: There is one example. This can be "Example**s**" when there are more.

File: docs/using/examples/example.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+certtemp.write(open(CA_FILE).read())
+certtemp.seek(0)
+ctx = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
+ctx.load_cert_chain(TEMP_CERT_CA_FILE, KEY_FILE)
+
+# Finally, create a HTTP connection.
+
+c = httplib.HTTPSConnection(CONTROL_SERVICE, CONTROL_PORT, context=ctx)
+
+def make_api_request(method, endpoint, data=None):
```
Comment (adamtheturtle) [inline docs/using/examples/example.py]: Let's meet our internal coding standards and best practices in our examples. Rather than enumerate them here, let's go through them together. A few guidelines are on http://doc-dev.clusterhq.com/gettinginvolved/contributing.html#merge-requirements but they are not complete.

File: docs/using/examples/api-example-python.rst [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+   * **CERT_FILE** - the API certificate file to use
+   * **CA_FILE** - the certificate authority file to use
+
+#. Now you can run the example:
+
+   .. prompt:: bash
+
+      python example.py
+
+.. _`Python 2.7.9`: https://www.python.org/downloads/
```
Comment (adamtheturtle) [inline docs/using/examples/api-example-python.rst]: Unfortunately instructing people on how to install Python is a game you can't win. Let's discuss how best to do this.

File: docs/using/examples/api-examples.rst (hunk 9 above)
Comment (davidoliverSP2) [inline docs/using/examples/api-examples.rst]: I think your concerns about adding these examples without automated testing needs to be considered by Roger and Michael, as these docs were requested for Dockercon, and Kai has put a lot of effort in to bring these in with manual tests. I'll message them directly.

**Comment by adamtheturtle** [comment]
Can we not use something to generate these examples? I understand that the ideal is to have human-crafted, idiomatic API requests in each language, but can we really do better than a tool? If so, is the difference worth the cost?

File: docs/using/examples/example.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
+import httplib
+import os
+import ssl
+import tempfile
+
+# NOTE - to run this example, you need at least python 2.7.9
+
+# Define the control IP, port, and the certificates for authentication.
+
+CONTROL_SERVICE = os.environ.get(
```
Comment (adamtheturtle) [inline docs/using/examples/example.py]: If we are going to go with manual testing (see other comments, and consider the results of our past attempts at doing manual testing for our docs / features) then it would be easier if there were a manual script to follow.

For example if we had a `flocker-provision` type tool which created a cluster which could be used with the default environment variables. Maybe hack `run-acceptance-tests` to help (I think it'll do most of the heavy lifting).

File: docs/using/examples/example.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
+import httplib
```
Comment (adamtheturtle) [inline docs/using/examples/example.py]: Buildbot should lint this file (file location can help).

**Comment by davidoliverSP2** [comment]
closed, following review by @ferrantim - this content is not required at this time.
